### ISSUE TITLE
show errors when failure happens

### ISSUE THREAD

**Issue by JulieGoldberg** [tobinbradley/dirt-simple-postgis-http-api#42]
If Dirt can't connect to the database, it should error in the console where it's running. There should be a verbose mode for seeing every request and potentially even more info. 

The issue I thought I had (https://github.com/tobinbradley/dirt-simple-postgis-http-api/issues/41) was because Dirt was returning some sort of error response, which Leaflet couldn't interpret as a tile. Googling Leaflet's error led me to think Dirt was compressing the tile and somehow headless browsers couldn't handle that. 

Logging would have avoided the rabbit hole I went down for days.

**Comment by tobinbradley** [comment]
Dirt does send route error messages back to the client. Unfortunately using a headless client makes those messages hard to discover.

I think what you're looking for is [Fastify's logging tools](https://www.fastify.io/docs/latest/Logging/). Fastify has logging turned off by default for performance reasons, but if you change line 3 of `index.js` to:

```javascript
const fastify = require('fastify')({ logger: true})
```

I think that'll get you what you're looking for.

**Comment by JulieGoldberg** [comment]
That gets me a list of requests -- but doesn't show me errors when it can't connect to the database when I explicitly put an invalid db connection in index.json. It still always produces a result with a status code of 200.

I'm looking for errors in the console where I'm running 'npm start' to run the tile server, not errors in puppeteer. 

If logging is off by default for performance, I'd suggest adding an areRequestsLogged parameter into index.json.txt that defaults to false (with a comment about why in the documentation) that gets used in index.js. That'll make it easier for others to turn it on for debugging purposes.

**Comment by tobinbradley** [comment]
If you look in the JSON that's returned when you can't connect to the database, you'll see this:

```javascript
{
  "statusCode": 500,
  "error": "Internal Server Error",
  "message": "unable to connect to database server"
}
```

That's pretty explicit. I can see how the 200 HTTP code could throw you off. While technically correct, a 200 code there isn't super helpful if you don't look at the API output itself. Your suggestion about making an option to toggle Fastify's logging is a good one. At the very least I could put a note about it in the readme file. 

Unfortunately if Fastify's logging doesn't cut it for you, you will have to manually stick `console.log()` all over the code base in any place you think the output might be interesting so it can write it to your console. That's not something I'd be on board with, but you can modify the code as you see fit. Dirt was designed to send that type of information to the client, the thought being (a) most people run Dirt in production in a way that they can't see the console, and (b) you may have lots of developers scattered across organizations and it's better for everybody involved if they can debug their own API calls without huddling around a terminal on a server someplace.

**Comment by JulieGoldberg** [comment]
Maybe there's something other than `const fastify = require('fastify')({ logger: true})` that I need to add to see the 500 errors. I'm not seeing them now.

I 100% agree that tons of console.log statements aren't ideal, and I do plan to run it in production without looking at the server output. But when setting it up, I definitely am looking at the console to try to ensure it's working. And having things written to logs is helpful if anything goes awry down the road in production.

**Comment by tobinbradley** [comment]
If you break your db connection and go to `http://localhost:3000/v1/list_tables` in your web browser, you'll see it. That's the JSON payload response from the server. 

It looks like normally Fastify sends a 500 code there, but since I sanitized some of the error output (the database connection error message can give away your server ip/port and database user name), Fastify was giving it a 200 code. I just pushed a patch that should give a 500 status code for DB connection issues, as well as a Fastify logger option in the configuration. You still won't get everything you're looking for in the console, but that should help you out.

Thanks for the feedback!
